### PR TITLE
Fix header space check

### PR DESCRIPTION
**Pull request by Ralith**
Fixes the failure identified in #2020.

### CODE REVIEW

File: quinn-proto/src/connection/mod.rs [hunk 1]
```diff
@@ -3739,8 +3748,24 @@ fn get_max_ack_delay(params: &TransportParameters) -> Duration {
...
+/// Excludes packet-type-specific fields such as packet number or Initial token
+// https://www.rfc-editor.org/rfc/rfc9000.html#name-0-rtt: flags + version + dcid len + dcid +
+// scid len + scid + length + pn
+const MAX_HANDSHAKE_OR_0RTT_HEADER_SIZE: usize =
+    1 + 4 + 1 + MAX_CID_SIZE + 1 + MAX_CID_SIZE + VarInt::from_u32(u16::MAX as u32).size() + 4;
+
+/// Minimal remaining size to allow packet coalescing, excluding cryptographic tag
+///
+/// This must be at least as large as the header for a well-formed empty packet to be coalesced,
+/// plus some space for frames. We only care about handshake headers because we short header packets
```
Comment (djc) [inline quinn-proto/src/connection/mod.rs:3763]: "we short header packets"?

File: quinn-proto/src/connection/packet_builder.rs [hunk 2]
```diff
@@ -138,12 +138,10 @@ impl PacketBuilder {
...
             )
         } else if space_id == SpaceId::Data {
-            let zero_rtt = conn.zero_rtt_crypto.as_ref().expect(
-                "sending packets in the application data space requires known 0-RTT or 1-RTT keys",
-            );
+            let zero_rtt = conn.zero_rtt_crypto.as_ref().unwrap();
             (zero_rtt.header.sample_size(), zero_rtt.packet.tag_len())
         } else {
-            unreachable!("tried to send {:?} packet without keys", space_id);
+            unreachable!();
```
Comment (djc) [inline quinn-proto/src/connection/packet_builder.rs:144]: Are these changes related? They don't look like improvements to me.

File: quinn-proto/src/connection/mod.rs [hunk 3]
```diff
@@ -3739,8 +3748,24 @@ fn get_max_ack_delay(params: &TransportParameters) -> Duration {
...
+/// necessarily have smaller headers, and initial packets are only ever the first packet in a
+/// datagram (because we coalesce in ascending packet space order and the only reason to split a
+/// packet is when packet space changes).
+const MIN_PACKET_SPACE: usize = MAX_HANDSHAKE_OR_0RTT_HEADER_SIZE + 32;
```
Comment (djc) [inline quinn-proto/src/connection/mod.rs:3767]: Nit: looks like we'd want to order this before `MAX_HANDSHAKE_OR_0RTT_HEADER_SIZE` rather than after?

**Review by djc — APPROVED** [review]

File: quinn-proto/src/connection/packet_builder.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/connection/packet_builder.rs:144]: The verbose panic messages were moved into `poll_transmit`, where they precede this logic.

Comment (djc) [reply quinn-proto/src/connection/packet_builder.rs:144]: Fair enough.

**Comment by Ralith** [comment]
Added another fix for a separate issue identified in a similar test.

**Review by djc — APPROVED** [review]
